### ISSUE TITLE
Multiple entry EntryProcessors do not respect locked entries

### ISSUE THREAD

**Issue by mmedenjak** [hazelcast/hazelcast#10065]
In case of multiple entry `EntryProcessor` (`MultipleEntryOperation` and `PartitionWideEntryOperation`), the processor will execute on the entry regardless if it is locked or not.

Currently the entire EntryProcessor is run on the partition thread so in theory we could check if all affected entries are locked and park the operation if so. On unlock & notification we re-check and again park the operation until all affected entries are unlocked. This sounds simple but would require a special `WaitNotifyKey` for a partition wide-notification and it would require some changes in the operation system since it now expects that one notification can notify a single waiting key.

With the introduction of EP `Offloadable` the situation is more complex since the EP can now be interleaved with other operations (also lock and unlock). This is why we need to check if the entry is locked before running the EP logic on every key. If the entry is locked, there are two possible approaches:

1. Park the EP operation. This requires substantial changes in the operation system since the operation isn't supposed to be parked/unparked after it started running.
2. Create a EntryOperation for the locked key and continue processing the unlocked keys. Once the EntryOperation is finished update the value on the multiple-entry operation and check if all entries have been processed. This is more in line with existing operations where parts of the work are forked onto other operations and using callbacks we notify the parent operation (e.g. in WAN - `WanSyncOperation`).

These tests demonstrate the difference introduced by adding `Offloadable` interface for EP which relies on locking keys for entry update.

```
@RunWith(HazelcastParallelClassRunner.class)
@Category({QuickTest.class, ParallelTest.class})
public class MixedEntryProcessorsTest extends HazelcastTestSupport {

    public static final String MAP_NAME = "EntryProcessorOffloadableTest";

    private HazelcastInstance[] instances;

    @Override
    public Config getConfig() {
        Config config = super.getConfig();
        MapConfig mapConfig = new MapConfig(MAP_NAME);
        mapConfig.setInMemoryFormat(OBJECT);
        mapConfig.setBackupCount(1);
        config.addMapConfig(mapConfig);
        return config;
    }

    @Before
    public void before() {
        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
        instances = factory.newInstances(getConfig());
    }

    @Test
    public void withOffloading() {
        runTest(true);
    }

    @Test
    public void withoutOffloading() {
        runTest(false);
    }

    private void runTest(final boolean offload) {
        final String key = generateKeyOwnedBy(instances[0]);
        final int initialMoney = 100000;
        User user = new User(initialMoney, 0, 0);

        final IMap<Object, Object> map = instances[1].getMap(MAP_NAME);
        map.put(key, user);
        int boughtItems = 0;

        for (int i = 0; i < 100; i++) {
            map.executeOnEntries(new BuyItem());
            boughtItems++;
        }

        final AtomicBoolean shouldRun = new AtomicBoolean(true);
        spawn(new Runnable() {
            @Override
            public void run() {
                map.executeOnKey(key, new CalculateBalance(offload));
                shouldRun.set(false);
            }
        });

        while (shouldRun.get()) {
            map.executeOnEntries(new BuyItem());
            boughtItems++;
        }

        map.executeOnKey(key, new CalculateBalance(offload));
        final User o = (User) map.get(key);
        System.out.println("CHECKING " + o);
        Assert.assertEquals(boughtItems, o.paidItems);
    }


    private static class User implements Serializable {

        public int money;
        public int itemsToPay;
        public int paidItems;

        User() {
        }

        public User(int money, int itemsToPay, int paidItems) {
            this.money = money;
            this.itemsToPay = itemsToPay;
            this.paidItems = paidItems;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (o == null || getClass() != o.getClass()) {
                return false;
            }

            User user = (User) o;

            if (money != user.money) {
                return false;
            }
            if (itemsToPay != user.itemsToPay) {
                return false;
            }
            return paidItems == user.paidItems;
        }

        @Override
        public int hashCode() {
            int result = money;
            result = 31 * result + itemsToPay;
            result = 31 * result + paidItems;
            return result;
        }

        @Override
        public String toString() {
            final StringBuffer sb = new StringBuffer("User{");
            sb.append("money=").append(money);
            sb.append(", itemsToPay=").append(itemsToPay);
            sb.append(", paidItems=").append(paidItems);
            sb.append('}');
            return sb.toString();
        }
    }

    private static class BuyItem implements EntryProcessor<String, User>, EntryBackupProcessor<String, User> {
        @Override
        public Object process(final Map.Entry<String, User> entry) {
            final User value = entry.getValue();
            value.itemsToPay++;
            entry.setValue(value);
            return value;
        }

        @Override
        public EntryBackupProcessor<String, User> getBackupProcessor() {
            return this;
        }

        @Override
        public void processBackup(Map.Entry<String, User> entry) {
            process(entry);
        }
    }

    private static class CalculateBalance implements EntryProcessor<String, User>,
            Offloadable,
            EntryBackupProcessor<String, User> {
        private boolean offload;

        public CalculateBalance() {
        }

        public CalculateBalance(boolean offload) {
            this.offload = offload;
        }

        @Override
        public Object process(final Map.Entry<String, User> entry) {
            // this is going to take a long time
            try {
                TimeUnit.SECONDS.sleep(1);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            final User value = entry.getValue();
            value.money -= value.itemsToPay;
            value.paidItems += value.itemsToPay;
            value.itemsToPay = 0;
            entry.setValue(value);
            return value;
        }

        @Override
        public EntryBackupProcessor<String, User> getBackupProcessor() {
            return this;
        }

        @Override
        public String getExecutorName() {
            return offload ? Offloadable.DEFAULT_ASYNC_EXECUTOR : Offloadable.NO_OFFLOADING;
        }

        @Override
        public void processBackup(Map.Entry<String, User> entry) {
            process(entry);
        }
    }

}
```

**Comment by mmedenjak** [comment]
Some more ideas on the subject :

> I had a very simple solution in mind for 3.9 (or maybe even 3.8.2)
> Create a LockAwareMapEntry with a method isLocked(), and the user may cast to this interface and check ifLocked in the EP running on multipleEntries.
> We can get info if entry locked or not very qucikly from lockStore since we're on partition thread anyway.
> 
> For 4.0 we can fix it fully - with two policies that we already had in mind:
> - either ignore locked entries and return to the user that these entries were not processed
> - or try locking and retry locking until succeeded or give up after retry-count
> These are changes we can't do in 3.x release stream, due to backward compatibilty.

**Comment by tombujok** [comment]
Thx, as I said in slack:
```
Thx for thinking about the EP locking - we can discuss it once I'm done with this story.
I had a very simple solution in mind for 3.9 (or maybe even 3.8.2)
Create a LockAwareMapEntry with a method isLocked(), and the user may cast to this interface and check ifLocked in the EP running on multipleEntries.
We can get info if entry locked or not very qucikly from lockStore since we're on partition thread anyway.

For 4.0 we can fix it fully - with two policies that we already had in mind:
- either ignore locked entries and return to the user that these entries were not processed
- or try locking and retry locking until succeeded or give up after retry-count
These are changes we can't do in 3.x release stream, due to backward compatibilty.
```

**Comment by tombujok** [comment]
```
I'm reluctant to implement it even for 3,9, since it will break people's code if they use map.lock and EP on entries.
but we can have a wider discussion - I'm sure we won't be able to implement anything in 3.8.x stream - the question is 3.9 vs. 4.0 -> but that we can discuss broader later on.
```

**Comment by mmedenjak** [comment]
This may be fixed without breaking changes if we allow configurable strategies. Moving this to 4.1.

**Comment by migralito** [comment]
I got REALLY confused about the fact that EntryProcessor javadoc states that entries passed to it will have the corresponding key locked. How is it possible that javadoc says something that it is not?!?!?!

> ...At the time you receive it the entry is locked and not released until the EntryProcessor completes...

**Comment by mmedenjak** [comment]
@migralito the javadoc might not quite be correct in its choice of words but in a way no-one should be able to change the entry while the `EntryProcessor` is running (unless you're using `com.hazelcast.core.Offloadable`). There is no actual locking involved.
This issue different. It is about entry processor methods on multiple entries (e.g. `IMap#executeOnEntries` and actual entry locks (e.g. `IMap#lock(K)`).

So the javadoc and this issue are about two different things.

**Comment by mmedenjak** [comment]
Related: https://github.com/hazelcast/hazelcast/issues/10935
